File: ios/Classes/Entities/CGMReading.swift
//
//  CGMReading.swift
//  basis_nfc_gm
//
//  Created by Michael Jajou on 1/27/23.
//

import Foundation

class CGMReading {
    var uid: String = ""
    var state: CGMState = .unknown
    var raw: String = ""
    var patchUID: String = ""
    var history: [UInt16] = []
    var trend: [UInt16] = []
    var minutesSinceStart: Int = 0
    
    init(
        uid: String = "",
        state: CGMState = .unknown,
        raw: String = "",
        patchUID: String = "",
        history: [UInt16] = [],
        trend: [UInt16] = [],
        minutesSinceStart: Int = 0
    ) {
        self.uid = uid
        self.state = state
        self.raw = raw
        self.patchUID = patchUID
        self.history = history
        self.trend = trend
        self.minutesSinceStart = 0
    }
}

enum CGMState: String, CaseIterable {
    case new
    case activating
    case unknown
    case operational
    var tagState: String { "Tag state: \(rawValue)" }
}

extension CGMReading {
    func toData() -> NSDictionary {
        return [
            "uid": uid,
            "state": state.rawValue.uppercased(),
            "raw": raw,
            "patchUID": patchUID,
            "history": history,
            "trend": trend,
            "minutesSinceStart": minutesSinceStart,
        ]
    }
}

File: lib/lib/cgm_reading.dart
// ignore_for_file: constant_identifier_names

part of cg_monitor;

class CGMReading {
  final String uid;
  final CGMState state;
  final String raw;
  final String patchUID;
  final List<int> history;
  final List<int> trend;
  final int minutesSinceStart;

  CGMReading({
    required this.uid,
    required this.state,
    required this.raw,
    required this.patchUID,
    required this.history,
    required this.trend,
    required this.minutesSinceStart,
  });

  factory CGMReading.fromJson(Map<String, dynamic> json) {
    return CGMReading(
      uid: json['uid'],
      state: CGMState.fromString(json['state']),
      raw: json['raw'],
      patchUID: json['patchUID'],
      history: json['history'],
      trend: json['trend'],
      minutesSinceStart: json['minutesSinceStart'],
    );
  }
}

enum CGMState {
  NEW,
  ACTIVATING,
  UNKNOWN,
  OPERATIONAL;

  static CGMState fromString(String val) {
    try {
      return CGMState.values.firstWhere((e) => e.name == val);
    } catch (_) {
      return CGMState.UNKNOWN;
    }
  }
}
